crates/fuel-core/src/state/historical_rocksdb.rs

```diff
@@ -295,20 +416,160 @@ where
+        // - A modification transaction M1 migrates height 42 from V1 to V2.
+        // - A second modification transaction M2 takes the cumulative changes from the lock, and adds the changes for another height.
+        // - A transaction T1 rollbacks to height 42, writes the changes to the modification history, and commits to rocksDB.
+        // - Transaction M2 acquires the lock to write the updated cumulative changes, which include the
+        //   migration of the history for height 42 from V1 to V2.
+        // - The next time a transaction commits to RocksDB, the cumulative changes from the migration transactions will be written to RocksDB,
+        //   which will cause the update from transaction T1 to be lost.
+
+        // This function is called only when a migration is in progress, hence the lock_guard is guaranteed to be Some.
+        let (cumulative_changes, lock_guard) = self.take_migration_changes();
+        let mut cumulative_transaction = StorageTransaction::transaction(
+            &self.db,
+            ConflictPolicy::Overwrite,
+            cumulative_changes,
+        );
+
+        // TODO: I could add a second migration transaction and commit the second transaction into the first
+        let mut migration_transaction = StorageTransaction::transaction(
+            &mut cumulative_transaction,
+            ConflictPolicy::Overwrite,
+            Changes::default(),
+        );
+
+        let v1_changes = migration_transaction
+            .storage_as_mut::<ModificationsHistoryV1<Description>>()
+            .take(&height)?;
+        if let Some(v1_changes) = v1_changes {
+            migration_transaction
+                .storage_as_mut::<ModificationsHistoryV2<Description>>()
+                .insert(&height, &v1_changes)?; // TODO: If this is an error we want to at least write back the old changes
+        };
+
+        // Add the changes into the cumulative changes
+        migration_transaction.commit()?;
+
+        self.add_migration_changes(lock_guard, cumulative_transaction.into_changes())?;
 
         Ok(())
     }
+
+    fn v1_entries(&self) -> BoxedIter<StorageResult<(u64, Changes)>> {
+        self.db
+            .iter_all::<ModificationsHistoryV1<Description>>(None)
+    }
+
+    fn is_migration_in_progress(&self) -> bool {
+        self.v1_entries().next().is_some()
+    }
+
+    #[cfg(test)]
+    fn commit_migration_changes(&self) -> StorageResult<()> {
+        self.db
+            .commit_changes(&std::mem::take(&mut *self.migration_changes.lock()))
+    }
+}
+
+// Try to take the value from `ModificationsHistoryV2` first.
+// If the migration is still in progress, remove the value from
+// `ModificationsHistoryV1` and return it if no value for `ModificationsHistoryV2`
+// was found. This is necessary to avoid scenarios where it is possible to
+// roll back twice to the same block height
+fn multiversion_take<Description, T>(
```

Comment:
I'd still suggest deduplicating these functions with:
```
fn multiversion_op<Description, T, F>(
    storage_transaction: &mut StorageTransaction<T>,
    height: u64,
    modifications_history_migration_in_progress: bool,
    f: F,
) -> StorageResult<Option<Changes>>
where
    Description: DatabaseDescription,
    T: KeyValueInspect<Column = Column<Description>>,
    F: FnOnce(
        StorageMut<'_, StorageTransaction<T>, ModificationsHistoryV2<Description>>,
    ) -> StorageResult<Option<Changes>>,
{
    let v2_last_changes =
        f(storage_transaction.storage_as_mut::<ModificationsHistoryV2<Description>>())?;

    if modifications_history_migration_in_progress {
        let v1_last_changes = storage_transaction
            .storage_as_mut::<ModificationsHistoryV1<Description>>()
            .take(&height)?;
        Ok(v2_last_changes.or(v1_last_changes))
    } else {
        Ok(v2_last_changes)
    }
}
```

Reply:
thanks, I need to become better at generic programming in Rust :) 
I tried to incorporate your changes from a previous PR of yours, but failed miserably while refactoring 😅

Reply:
Added your change: 5ea1bbcb0c668d7d49aeb049001a8f3648bee6e0